Split cells via Min Cut

philippotto opened the pull request:
This PR:
- implements a min-cut approach to cut a cell given two seeds
- fixes a rendering bug where fallback rendering did not use the best available mag sometimes

### URL of deployed dev instance (used for testing):
- https://mincut.webknossos.xyz

### Steps to test:
- open a volume tracing with fallback layer
- use adhoc meshing for a cell to find a good spot to cut (e.g., a spinehead)
- create a tree place two connected nodes in the cell which represent the seeds (e.g., one in the dendrite and one in the spinehead)
- rightclick the tree and select "perform min-cut"
- a min cut should be performed which erases some voxels between the two seeds
- you can verify that the split was successful by using the floodfill with a new segment id on one half of the segment

### To do:
- [X] implement working version
- [X] clean up UI
- [X] clean up code
- [x] pull master
- [x] generalize mag handling (currently, the existence of 2-2-1 is assumed)
- [x] try to refine min-cut on better mags
- [x] make operation undo-able
- ~[ ] fix flicker bug when zooming from mag 4 to mag 8.~ extracted to https://github.com/scalableminds/webknossos/pull/6029

### Issues:
- fixes #5934 

------
(Please delete unneded items, merge only when none are left open)
- [x] Updated [(unreleased) changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [(unreleased) migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-connect](https://github.com/scalableminds/webknossos-connect) if datastore API changes
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change
- [ ] Needs datastore update after deployment
- [x] Ready for review

philippotto commented on .circleci/not-on-master.sh:
```diff
@@ -1,8 +1,8 @@
 set -Eeuo pipefail
 
-if [ "${CIRCLE_BRANCH}" == "master" ]; then
+# if [ "${CIRCLE_BRANCH}" == "master" ]; then
```
todo

philippotto commented on frontend/javascripts/oxalis/api/api_latest.js:
```diff
@@ -1233,16 +1233,16 @@ class DataApi {
     }
 
-    const { resolutions } = layer;
+    const resolutions = resolutionInfo.getDenseResolutions();
```
this is a bug fix.

philippotto commented on frontend/javascripts/oxalis/default_state.js:
```diff
@@ -95,7 +95,7 @@ const defaultState: OxalisState = {
     },
     preferredQualityForMeshPrecomputation: 2,
-    preferredQualityForMeshAdHocComputation: 2,
+    preferredQualityForMeshAdHocComputation: 1,
```
todo: revert

philippotto commented on frontend/javascripts/oxalis/controller/scene_controller.js:
```diff
@@ -136,6 +136,22 @@ class SceneController {
       return cube;
     };
 
+    window.addVoxelMesh = (position: Vector3, cubeLength: Vector3, optColor?: string) => {
```
I used this for debugging purposes, but there seems to be a slight offset in the position. Not sure why, though.

philippotto replied:
Ah, now the penny has dropped :smile: However, the `0.9` is not fully responsible because I saw the offset already before I added the 0.9 factor. As you correctly point out, the flooring is messing things up. Since I used a cube length of `[1, 1, 1]`, the division by two (which is necessary since threejs expects the position of the cube to be in the center) produced fractional numbers which were important for rendering at the correct position, but were removed again by the flooring. I don't know why I even added the flooring in the first place. Maybe I thought, that there would be problems with fractionals :shrug:
I removed the flooring and also explained the 0.9 factor (which I added so that one can see the border between voxels in form of a gap) which wasn't used when instantiating the box geometry, by the way. I guess, it was late when I added this code :sweat_smile:

philippotto commented on frontend/javascripts/oxalis/model/bucket_data_handling/texture_bucket_manager.js:
```diff
@@ -388,9 +388,16 @@ export default class TextureBucketManager {
+          ) {
+            // Another bucket was already placed here with a better zoomstep.
+            // Skip the entire loop.
+            // TODO: clarify flicker when zooming from mag 4 to mag 8.
```
todo: I might need a rubber duck here :see_no_evil:

philippotto replied:
I reverted the changes and extracted them to a separate PR: https://github.com/scalableminds/webknossos/pull/6029. Since this proved to be a bit tricky, a separate PR is easier to revert if it turns out to cause trouble in production. Also, it doesn't block the current PR then.

philippotto commented:
@daniel-wer I think, this PR is ready for review. Let me know whether you want a high-level overview of this PR before you dive into it :) I tried to add many comments, but I'm not sure how well the big picture is explained.

philippotto commented on frontend/javascripts/oxalis/model/sagas/min_cut_saga.js:
```diff
@@ -0,0 +1,694 @@
+  const globalSeedA = V3.fromMag1ToMag(nodes[0].position, targetMag);
+  const globalSeedB = V3.fromMag1ToMag(nodes[1].position, targetMag);
+
+  let minDistToSeed = MIN_DIST_TO_SEED / targetMag[0];
```
this is not correct, since the three dimensions have to be taken into account each.

philippotto commented on frontend/javascripts/oxalis/model/sagas/min_cut_saga.js:
```diff
@@ -0,0 +1,694 @@
+  const ll = ([x, y, z]) => z * size[1] * size[0] + y * size[0] + x;
+
+  if (inputData[ll(seedA)] !== inputData[ll(seedB)]) {
+    console.warn(
```
toast

philippotto commented on frontend/javascripts/oxalis/model/sagas/min_cut_saga.js:
```diff
@@ -0,0 +1,694 @@
+  const edgeBuffer = new Uint16Array(length);
+  for (let x = 0; x < size[0]; x++) {
+    if (x % Math.floor(size[0] / 10) === 0 && performance.now() > timeoutThreshold) {
+      // After each 10% chunk, we yield to redux-saga to allow
+      // cancellation.
```
```suggestion
      // After each 10% chunk, we check whether there's a timeout
```

philippotto commented on frontend/javascripts/oxalis/model/sagas/min_cut_saga.js:
```diff
@@ -0,0 +1,694 @@
+      // cancellation.
+      throw TimeoutError;
+    }
+    for (let y = 0; y < size[1]; y++) {
+      for (let z = 0; z < size[2]; z++) {
```
should I reorder the x-y-z loops to z-y-x for better cache locality? probably no harm in changing it.

philippotto commented on frontend/javascripts/oxalis/model/sagas/min_cut_saga.js:
```diff
@@ -0,0 +1,694 @@
+    const currentDistance = distanceField[ll(currentVoxel)];
+
+    if (V3.equals(currentVoxel, seedA)) {
+      // console.log("Finished removing shortest path. Deleted edges:", removedEdgeCount);
```
```suggestion
```

philippotto commented on frontend/javascripts/oxalis/model/sagas/min_cut_saga.js:
```diff
@@ -0,0 +1,694 @@
+    const currDist = distanceField[ll(neighborPos)];
+    const distToSeed = Math.min(currDist, maxDistance - currDist);
+
+    if (distToSeed > minDistToSeed) {
```
```suggestion
    // We don't want to remove voxels which are close to the seeds.
    // See explanation for MIN_DIST_TO_SEED for details.
    if (distToSeed > minDistToSeed) {
```

daniel-wer commented on frontend/javascripts/oxalis/model/sagas/min_cut_saga.js:
```diff
@@ -0,0 +1,722 @@
+    }
+
+    if (!(distanceField[ll(neighborPos)] < currentDistance && distanceField[ll(neighborPos)] > 0)) {
+      throw new Error("Direction points towards a higher distance?");
```
```suggestion
      throw new Error("Direction points towards an equal/higher/zero distance?");
```

daniel-wer commented on frontend/javascripts/oxalis/model/sagas/min_cut_saga.js:
```diff
@@ -0,0 +1,722 @@
+  // Perform a breadth-first search from seedA and return
+  // which voxels were visited (visitedField). This represents
+  // the graph component for seedA.
+  const visitedField = new Uint16Array(boundingBoxTarget.getVolume());
```
Since the field only holds 0s and 1s, a `Uint8Array` is sufficient.
```suggestion
  const visitedField = new Uint8Array(boundingBoxTarget.getVolume());
```

daniel-wer commented on frontend/javascripts/oxalis/model/sagas/min_cut_saga.js:
```diff
@@ -0,0 +1,722 @@
+  // which voxels were visited (visitedField). This represents
+  // the graph component for seedA.
+  const visitedField = new Uint16Array(boundingBoxTarget.getVolume());
+  const queue: Array<{ voxel: Vector3 }> = [{ voxel: seedA }];
```
Could also simply be `Array<Vector3>`

daniel-wer commented on frontend/javascripts/oxalis/model/sagas/min_cut_saga.js:
```diff
@@ -0,0 +1,722 @@
+                V3.add(boundingBoxTarget.min, neighborPos),
+                targetMag,
+              );
+              for (let dx = 0; dx < targetMag[0]; dx++) {
+                for (let dy = 0; dy < targetMag[1]; dy++) {
+                  for (let dz = 0; dz < targetMag[2]; dz++) {
+                    api.data.labelVoxels([V3.add(position, [dx, dy, dz])], 0);
+                  }
+                }
+              }
```
Wouldn't it be equivalent to label the single voxel in the target mag and let the upsampling take care of the rest. Or is it needed that the voxels in mag 1 are immediately labeled?

daniel-wer commented on frontend/javascripts/oxalis/model/sagas/min_cut_saga.js:
```diff
@@ -0,0 +1,722 @@
+  yield* takeEveryUnlessBusy("PERFORM_MIN_CUT", performMinCut, "Min-cut is being computed.");
+}
+
+// Todo: remove before merging
```
TODO

philippotto commented on frontend/javascripts/oxalis/model/sagas/min_cut_saga.js:
```diff
@@ -0,0 +1,722 @@
+function _getNeighborsFromBitMask(bitMask) {
+  // Note: Use the memoized version of this: getNeighborsFromBitMask.
+  // Ingoing and outgoing edges are stored as a bitmask. The first half
+  // of the bitmask holds the ingoing edges as bits. The second half the
```
Probably depends on how one interpets "first half" :D I meant first half when viewing the binary representation (going from high-value to low-value). However, I'll simply rephrase it to be more precise.

philippotto commented on frontend/javascripts/oxalis/view/context_menu.js:
```diff
@@ -295,18 +309,20 @@ function NodeContextMenuOptions({
-      >
-        Path Length to this Node
-      </Menu.Item>
+      {isTheSameNode ? null : (
```
It bugged me that rightclicking a selected node always showed "Path length to this node: 0 vx".

philippotto commented on frontend/javascripts/oxalis/model/sagas/saga_helpers.js:
```diff
@@ -0,0 +1,41 @@
+
+    yield* put(setBusyBlockingInfoAction(true, reason));
+    yield* call(saga, action);
+    yield* put(setBusyBlockingInfoAction(false));
```
The busy-info stuff is meant as a sort of mutex. I just checked all occurrences (there are not many) and all use the pattern correctly. As a safe guard, I added an assertion to the corresponding reducer so that an error is thrown in case the mutex is not handled properly (now or in the future). See 24acd467b.

philippotto commented on frontend/javascripts/oxalis/model/sagas/min_cut_saga.js:
```diff
@@ -0,0 +1,722 @@
+  } else if (boundingBoxes.length === 1) {
+    boundingBoxObj = boundingBoxes[0].boundingBox;
+  } else {
+    console.log(
```
> My gut tells me that this bounding box business is a little finicky and opaque. I'm not sure whether users will understand the effect of having bounding boxes visible, since it doesn't seem to be communicated. I can also imagine that it is somewhat annoying having to disable all (but one) bounding boxes every time a min-cut should be performed OR users forgetting they have an unrelated bounding box visible.

You are absolutely right. This was bugging me, too. I now changed the code so that one can choose which bounding box should be used from the context menu. Only the bounding boxes which contain the seed nodes are shown in the list. Additionally, one can select "Create new bounding box". I'm quite happy with the UX now :)
The sad news is that this change required to use antd submenus within the context menu which turned out to be very painful. I had to dig quite deep and tried several approaches to get this working. The final solution can be seen in the following commits. The link points to a comment where I explained the rationale behind this. I suggest to review these commits separately. Also note that I had to move quite a lot of code within the context menu module. Most of it stayed as is, but I added a new `<Dropdown />` component and separated it from the actual container. The upside is that I could remove quite a lot of custom CSS classes which we passed to antd menus :tada: 

https://github.com/scalableminds/webknossos/pull/5885/files/600dc941fe5e5a251cba60112b23fdb3fc846f78..58562dd90b1c73bc35b35cafe54215ee04fdf56d#diff-989748590830a0e73eef9fd9e98df428dab5559c0a88e4a656e9bd2814f3cb65R669

> Was the customizable bounding box a requirement for this feature? Otherwise, we could start by always using the automatically sized bounding box and think about adding a padding size setting or a bounding box selector once it is requested thinking

This would have been alright for the first iteration probably, but when developing the feature I quickly noticed that a poor bounding box can lead to poor min cut results. Ideally, it should cover exactly the merger (not too much and not too little).  That's why I wanted to make it easy to select the bbox. If I would have known that this was going to be so tricky, I might have decided differently :sweat_smile: However, I'm confident that this context menu re-architecturing will be helpful in the future, too.

philippotto commented on frontend/javascripts/oxalis/model/sagas/min_cut_saga.js:
```diff
@@ -0,0 +1,722 @@
+
+// To choose the initial mag, a voxel threshold is defined
+// as a heuristic. This avoids that an unrealistic mag
+// is tried in the first place.
```
> Or does this result in worse cuts?

Yes, I think, when taken to the extreme (think of mag 1024), the "fat erasing" will definitely cause harm as it can easily erase too much. Still, there are probably ways to improve the heuristic, but I think for the first version, this should be fine.

philippotto commented on frontend/javascripts/libs/mjs.js:
```diff
@@ -209,6 +209,10 @@ V3.divide3 = function divide3(a, k, r) {
   return r;
 };
 
+V3.fromMag1ToMag = (vec, targetMag) => V3.floor(V3.divide3(vec, targetMag));
+
+V3.fromMagToMag1 = (vec, sourceMag) => V3.floor(V3.scale3(vec, sourceMag));
```
I found it appealing to put it here, since the V3 module contains all operations which we perform on Vector3's. I know, that we haven't really used the domain knowledge of mags/resolutions here, but I think, it makes sense to do so. Overall, the entire mag/resolution handling (starting with the naming) could use a good cup of refactoring, obviously.. :see_no_evil:

philippotto commented:
@daniel-wer I think, I covered all your feedback and the remaining to dos :) Feel free to test again and have a look at the code. Also, see my one comment about the rationale behind the context menu changes: https://github.com/scalableminds/webknossos/pull/5885#discussion_r802808843

daniel-wer reviewed (commented):
Nice improvements! Looking forward to testing the reworked bounding box selection, I'll report back, afterwards :)

daniel-wer commented:
My notes from testing:

- [x] The bounding box selection and "happy path" (meaning nodes with a somewhat thin connection between them) work really well :partying_face: 
- [ ] Since the min-cut can take a while (seconds to minutes), some kind of progress indicator is needed in my opinion. Could simply be antd `Message`s showing the group descriptions that you are already using in the console (via `progress_callback.js`).
- [ ] I had a case that took a little longer and the console showed that "Find & delete paths" took more than 20s. I would've expected it to switch to the next-higher mag after 10 seconds as indicated in the code. The dataset had higher resolutions.
  ```
  Trying min-cut computation at 2-2-1
    Setting minDistToSeed to  12.5
    Loading data... (for 439350 vx)
    Starting min-cut...
    Build Graph: 65.99365234375 ms
    Segmentation is partitioned
    Find & delete paths: 21472.55126953125 ms
    traverseResidualsField: 38.225830078125 ms
    labelDeletedEdges: 137.98681640625 ms
    Total min-cut: 21714.920166015625 ms
  Refining min-cut at 1-1-1
    ...
  ```
- [ ] I had a min-cut reaching the bounding box border, because the cell was not entirely contained, leading to a spherical "min-cut" of one of the nodes. I understand why this happens and that it's not easy to solve, but maybe we could notify the user that the bounding box was likely too small if the edge(s) is/are reached. What do you think?
- [ ] [Low-Prio] I tried to min-cut a thicker cell which triggered the switch to the next-higher mag (which is good :+1:). In that case, some of the timers are not reset correctly, it seems
  ```
  min_cut_saga.js:404 Timer 'Total min-cut' already exists
  min_cut_saga.js:422 Timer 'Find & delete paths' already exists
  ```

philippotto commented:
Thanks for your feedback, @daniel-wer! I think, I incorporated almost all of it. Would you mind having another look at my newest commits?

> I had a case that took a little longer and the console showed that "Find & delete paths" took more than 20s. I would've expected it to switch to the next-higher mag after 10 seconds as indicated in the code.

I could reproduce this :+1: One modulo operation used a quite high operand which I changed from 100000 to 10000. Now, it should work better.

> I had a min-cut reaching the bounding box border, because the cell was not entirely contained, leading to a spherical "min-cut" of one of the nodes. I understand why this happens and that it's not easy to solve, but maybe we could notify the user that the bounding box was likely too small if the edge(s) is/are reached. What do you think?

This is the only thing I skipped now. I'd like to get some feedback from real-world usages before I fune-tune this feature further. The feature itself is marked as "experimental" in the UI, anyway, and there might be other more impactful levers later which we need to turn. I hope this is alright with you? :)

daniel-wer reviewed (approved):
Code LGTM :partying_face: 

> This is the only thing I skipped now. I'd like to get some feedback from real-world usages before I fune-tune this feature further. The feature itself is marked as "experimental" in the UI, anyway, and there might be other more impactful levers later which we need to turn. I hope this is alright with you? :)

Absolutely, makes sense :)
